proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -314,49 +279,210 @@ protocol TargetBuildContext {
         var path: Path { get }
     }
 }
+```
 
-/// Constructs commands to run during the build, including command lines,
-/// environment variables, initial working directory, etc. All paths should
-/// be based on the ones passed to the plugin in the target build context.
-protocol CommandConstructor {
+##### Package Model
```

Comment:
should the model all really be all protocols and not structs? do we think there will be multiple implementation of them?

Reply:
That's a good point.  In the original I wanted to make them protocols to focus on the API and not implementation, but it makes more sense to show them as they are.

I wanted to make them classes, actually, so that we can use specialization to represent targets and products.  This is more open-ended than using enums, which I think will both scale less well as new kinds of products are added and also be less ergonomic than being able to do, say `targets.filter { $0 is BinaryLibraryTarget }`.  Using classes is also more conceptually correct since the types have identity.

Switching away from protocols also makes it possible to nest structs in the declarations where appropriate.  I'll update the declarations and we can see how it looks.